packages/cli/src/commands/test-v2.ts

```diff
@@ -36,35 +92,199 @@ const commandModule: CommandModule<Options, Options> = {
-    process.exit(0);
+    const signals: { testsChanged: string[]; sourceChanged: string[]; configChanged: boolean } = {
+      testsChanged: [],
+      sourceChanged: [],
+      configChanged: false,
+    };
+
+    function resetSignals() {
+      signals.testsChanged = [];
+      signals.sourceChanged = [];
+      signals.configChanged = false;
+    }
+
+    const signalEvents = new EventEmitter();
+
+    chokidar.watch([configPath, testDir, srcDir]).on("all", async (_, path) => {
+      if (path.includes(testDir)) {
+        signals.testsChanged.push(path);
+        signalEvents.emit(Events.TESTS_CHANGED, signals.testsChanged);
+      } else if (path.includes(srcDir)) {
+        signals.sourceChanged.push(path);
+        signalEvents.emit(Events.SOURCES_CHANGED, signals.sourceChanged);
+      } else if (path.includes(configPath)) {
+        signals.configChanged = true;
+        signalEvents.emit(Events.CONFIG_CHANGED, path);
+      }
+    });
+    await new Promise((resolve) => setTimeout(resolve, 100));
+    resetSignals();
+
+    try {
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
+        await runCodegen();
+        const worldAddress = await deployWorld();
+        console.log(chalk.blue("World address", worldAddress));
+        // Create a temporary file to pass the world address to the tests
+        writeFileSync(WORLD_ADDRESS_FILE, worldAddress);
+
+        // eslint-disable-next-line no-constant-condition
+        while (true) {
+          try {
+            await runTests();
+          } catch (e) {
+            if (e instanceof CommandFailedError) {
+              console.error(chalk.red(chalk.bold("Tests failed")));
+              if (verbosity < 3) {
+                console.log(
+                  chalk.yellow('To get additional traces for your tests, set "verbosity" in your foundry.toml to "3"')
+                );
+              }
+            } else {
+              console.error(e);
+              break;
+            }
+          }
+
+          if (signals.configChanged || signals.sourceChanged.length > 0 || signals.testsChanged.length > 0) {
+            if (signals.configChanged) {
+              console.log(chalk.green("Config changed. redeploying world"));
+              resetSignals();
+              break;
+            } else if (signals.sourceChanged.length > 0) {
+              console.log(chalk.green("System changed. redeploying world"));
+              for (const filesChanged of signals.sourceChanged) {
+                console.log(chalk.gray("changed: ", filesChanged));
+              }
+              resetSignals();
+              break;
+            } else {
+              console.log(chalk.green("Tests changed. rerunning forge tests."));
+              for (const filesChanged of signals.testsChanged) {
+                console.log(chalk.gray("changed: ", filesChanged));
+              }
+              resetSignals();
+              continue;
+            }
+          } else {
```

Comment:
seems like you could reduce nesting depth by removing the first condition and doing `else if (signals.testsChanged.length > 0)` and then `else`